### PR TITLE
Error and Reporting Alignment : Phase 1

### PR DESCRIPTION
**Pull request by shellscape**
**What kind of change does this PR introduce?**

Refactor, RFC Implementation

**Did you add tests for your changes?**

Not yet needed, but tests were updated to satisfy changes in this PR.

**If relevant, link to documentation update:**

Not yet needed, but will occur when the validation and deprecation messages are enabled.

**Summary**

This PR is Phase 1 for changes proposed in RFC #6534. It introduces `WebpackErrorArray` and implemented a refactor to `WebpackError`. This is setup required for future work. The use of Errors across the webpack codebase is inconsistent and spread out. Many tests rely on arbitrary tests that differ greatly. It's for those reasons that a phased approach to fulfill the RFC is required.

To be implemented in later phases to follow shortly:
- Create proper Error types which extend `WebpackError` for code using `new Error`
- Enforcement in `WebpackError` construction
- Enforcement of types pushed to `WebpackErrorArray`
- Standardization of `toString` for `WebpackError`

**Does this PR introduce a breaking change?**

Nope

**Other information**

### CODE REVIEW

File: lib/WebpackError.js [hunk 1]
```diff
@@ -4,8 +4,81 @@
...
+	static deprecate(message) {
+		const err = new Error();
+		const stack = err.stack
+			.split("\n")
+			.slice(3, 4)
+			.join("\n");
+		// use process.stdout to assert the message will be displayed in
+		// environments where console has been proxied. this mimics node's
+		// util.deprecate method.
+		process.stdout.write(`DeprecationWarning: ${message}\n${stack}\n`);
```
Comment (ooflorent) [inline lib/WebpackError.js]: Why not using `process.emitWarning`? That's what is used by `util.deprecated`.
https://github.com/nodejs/node/blob/master/lib/internal/util.js#L60

File: lib/WebpackError.js [hunk 2]
```diff
@@ -4,8 +4,81 @@
...
+		// }
+		// TODO: enable message length limit at a commit to follow
+		// else if (message.length > 160) {
+		// 	WebpackError.deprecate("message cannot be longer than 80 characters.");
+		// }
+
+		super(message);
+
+		this[hash] = createHash("md4");
+		this[meta] = {};
```
Comment (ooflorent) [inline lib/WebpackError.js]: `Object.create(null)` or `{__proto__: null}` would be better

File: lib/WebpackError.js [hunk 3]
```diff
@@ -4,8 +4,81 @@
 */
 "use strict";
 
+const createHash = require("./util/createHash");
+
+const hash = Symbol();
+const meta = Symbol();
```
Comment (ooflorent) [inline lib/WebpackError.js]: Could you name the symbols?

Comment (shellscape) [reply lib/WebpackError.js]: For what purpose? There's no best practice outlining that should be done, and these are not meant to be used outside of the class.

File: lib/WebpackError.js (hunk 2 above)
Comment (shellscape) [reply lib/WebpackError.js]: Do you have documentation I can review showing why those two alternatives are better than `{}`?

File: lib/WebpackError.js (hunk 1 above)
Comment (shellscape) [reply lib/WebpackError.js]: I found the `(node nnnn)` prefix to be superfluous in the log. It doesn't add any value, hence the util function didn't add any value.

Comment (sokra) [reply lib/WebpackError.js]: Yep that's important for `--no-deprecation` etc to work

File: lib/WebpackError.js [hunk 4]
```diff
@@ -4,8 +4,81 @@
...
+
+	get id() {
+		return this[hash];
+	}
+
 	inspect() {
 		return this.stack + (this.details ? `\n${this.details}` : "");
 	}
+
+	get meta() {
```
Comment (sokra) [inline lib/WebpackError.js]: Why is this needed?

File: lib/WebpackError.js [hunk 5]
```diff
@@ -4,8 +4,81 @@
...
+			.split("\n")
+			.slice(3, 4)
+			.join("\n");
+		// use process.stdout to assert the message will be displayed in
+		// environments where console has been proxied. this mimics node's
+		// util.deprecate method.
+		process.stdout.write(`DeprecationWarning: ${message}\n${stack}\n`);
+	}
+
+	get id() {
```
Comment (sokra) [inline lib/WebpackError.js]: Why is this needed?

File: lib/WebpackError.js [hunk 6]
```diff
@@ -4,8 +4,81 @@
...
 
+const createHash = require("./util/createHash");
+
+const hash = Symbol();
+const meta = Symbol();
+
 module.exports = class WebpackError extends Error {
+	constructor(
+		message,
+		line = 0,
```
Comment (sokra) [inline lib/WebpackError.js]: `lineEnd` would make sense for multiple line pointers.

The parser uses location objects in this format: `{ start: { line, column }, end: { line, column } }`. Does it make sense to use the same format here instead of multiple arguments?

File: lib/WebpackError.js [hunk 7]
```diff
@@ -4,8 +4,81 @@
...
+		column = 0,
+		columnEnd = 0,
+		category = "error"
```
Comment (sokra) [inline lib/WebpackError.js]: Is this redundant information? It's already stated by the name of the array to which the error is pushed to.

Can I push an error to the `compilation.warnings` array?

File: lib/WebpackError.js (hunk 2 above)
Comment (ooflorent) [reply lib/WebpackError.js]: No I don't have one. The purpose of these 2 alternatives is to have an object with a `null` prototype. Lookups on it would be faster since it doesn't require traversing the proto chain.

File: lib/WebpackError.js (hunk 4 above)
Comment (shellscape) [reply lib/WebpackError.js]: This is part of the RFC. Extraneous values (such as file lists) which are currently placed into an error's message (making them egregiously large) will be placed into the meta property. Error types that derive from WebpackError can determine how best to use the meta property data.

File: lib/WebpackError.js (hunk 5 above)
Comment (shellscape) [reply lib/WebpackError.js]: This will be important and useful for reporters in the future.

File: lib/WebpackError.js (hunk 3 above)
Comment (ooflorent) [reply lib/WebpackError.js]: This is for debugging purpose. Here below, you can easy identify what `b` is whereas `a` is not.

symbols

File: lib/WebpackError.js (hunk 6 above)
Comment (shellscape) [reply lib/WebpackError.js]: I have no issue with using the same format. The only reason I chose this format is that it happens to be the same that Firefox uses. Otherwise, there is no standard for this.

File: lib/WebpackError.js (hunk 1 above)
Comment (ooflorent) [reply lib/WebpackError.js]: This flag sets `process.noDeprecation` which is handled by `util.deprecated`.

File: lib/WebpackError.js (hunk 7 above)
Comment (shellscape) [reply lib/WebpackError.js]: It is not. Category will be used in further PRs, which satisfies a portion of the RFC allowing errors to use a name/category for identifying what the error is for. For example, configuration errors would not include "configuration" as a prefix, rather they would set the category to "configuration" and keep the error message property short. "error", "warning", "configuration", "performance" are all examples of categories. This will be utilized more when the abstract enforcement goes into place.

File: lib/WebpackError.js (hunk 1 above)
Comment (sokra) [reply lib/WebpackError.js]: https://nodejs.org/api/process.html#process_process_emitwarning_warning_type_code_ctor

> The following additional handling is implemented if the warning type is DeprecationWarning:
> *   If the --throw-deprecation command-line flag is used, the deprecation warning is thrown as an exception rather than being emitted as an event.
> *   If the --no-deprecation command-line flag is used, the deprecation warning is suppressed.
> *   If the --trace-deprecation command-line flag is used, the deprecation warning is printed to stderr along with the full stack trace.

Seems like there is some magic in emitWarning... ^^

Comment (ooflorent) [reply lib/WebpackError.js]: Since `util.deprecate` checks for `process.noDeprecation` I didn't look at `emitWarning` implementation. Then better use `emitWarning` I guess.

Comment (shellscape) [reply lib/WebpackError.js]: OK are we on the same page that `process.emitWarning` is preferred here then?

File: lib/WebpackError.js (hunk 2 above)
Comment (shellscape) [reply lib/WebpackError.js]: OK. Small gains are small gains. I'll update that.

File: lib/WebpackError.js (hunk 3 above)
Comment (shellscape) [reply lib/WebpackError.js]: Fair enough. I'll update this.

File: lib/WebpackError.js [hunk 8]
```diff
@@ -4,8 +4,83 @@
...
 
+const createHash = require("./util/createHash");
+
+const hash = Symbol("hash");
+const meta = Symbol("meta");
+
 module.exports = class WebpackError extends Error {
+	constructor(
+		message,
+		options = Object.create({
```
Comment (ooflorent) [inline lib/WebpackError.js]: Why did you use `Object.create` instead of a plain object?

File: lib/WebpackError.js [hunk 9]
```diff
@@ -4,8 +4,83 @@
...
+		// }
+	}
+
+	static deprecate(message) {
+		const err = new Error();
+		const stack = err.stack
+			.split("\n")
+			.slice(3, 4)
+			.join("\n");
+		// use process.stdout to assert the message will be displayed in
```
Comment (ooflorent) [inline lib/WebpackError.js]: Comment is outdated

File: lib/WebpackErrorArray.js [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+	push(...args) {
+		// TODO: validation will be enabled in a later PR
+		// WebpackErrorArray.validate(...args);
+		super.push(...args);
+	}
+
+	static validate(...args) {
+		for (const arg of args) {
+			if (!(arg instanceof WebpackError)) {
```
Comment (ooflorent) [inline lib/WebpackErrorArray.js]: This will be a long transition to plugin authors. We must be prepared to receive a lot of issues about this message.

File: lib/WebpackError.js (hunk 8 above)
Comment (shellscape) [reply lib/WebpackError.js]: For the same reasons you stated on your prior request?

File: lib/WebpackError.js (hunk 9 above)
Comment (shellscape) [reply lib/WebpackError.js]: Lord have mercy

File: lib/WebpackErrorArray.js (hunk 10 above)
Comment (shellscape) [reply lib/WebpackErrorArray.js]: Yep. Growing is hard. But if you've seen the garbage that plugin authors are pushing to the errors and warnings arrays... It'll be worth it.

File: lib/WebpackError.js (hunk 8 above)
Comment (ooflorent) [reply lib/WebpackError.js]: I talked about `Object.create(null)` to create an object with an empty prototype. In this case, it creates an object with `{…}` as prototype. Here, a plain object is better ^^

File: lib/WebpackErrorArray.js (hunk 10 above)
Comment (ooflorent) [reply lib/WebpackErrorArray.js]: I have no idea about that! Could you share your _discoveries_? It may be a good idea to brew around what they are trying to do and how to provide a better DX.

Comment (shellscape) [reply lib/WebpackErrorArray.js]: I sure can. Most of that was discovered while working on `webpack-stylish`. I'll add examples of those to the RFC thread this week for reference. There's already a lot of code in `webpack-stylish` that attempts to deal with the different kinds of data (strings, arrays, Errors) that get added to those collections. One more reason locking this down will really help the project and ecosystem in the future.

File: lib/WebpackError.js (hunk 8 above)
Comment (shellscape) [reply lib/WebpackError.js]: Fixed!

File: lib/WebpackError.js (hunk 9 above)
Comment (shellscape) [reply lib/WebpackError.js]: Fixed!

**Comment by shellscape** [comment]
As mentioned off-Github: This PR is being blocked based on commented-out code within the PR, despite the fact that no guidelines exist within CONTRIBUTING.md which explicitly forbid that practice, and despite the fact that the comments are pertinent to quick-to-follow phases related to and reliant upon the code within the `//TODO:` blocks, and that the commented-out code is expressive and speaks to intent and guidance for changes to follow.

I understand that disagreements with a maintainer's choices are inevitable, but I disagree with applying arbitrary rules to Pull Requests on a whim. For these reasons I'm closing this PR and abandoning my effort to improve webpack's `Error` use. This is a net loss for the user experience.
